File: day07.nim
import strutils, tables

const instructions = readFile("./inputs/07.txt").strip.splitlines

type
  Wire = ref object
    value: int
    hasValue: bool
    logic: seq[string]


proc initialize(): Table[string, Wire] =
  result = initTable[string, Wire]()
  for line in instructions:
    let values = line.split(" -> ")
    result[values[1]] = Wire(logic: values[0].split())

var wires = initialize()


proc getValue(wireName: string): int =
  if not wires.hasKey(wireName): return wireName.parseInt()
  if wires[wireName].hasValue: return wires[wireName].value

  let
    wire = wires[wireName]
    logic = wire.logic

  case logic.len
    of 1: result = getValue(logic[0])
    of 2: result = not getValue(logic[1])
    else:
      case logic[1]
        of "AND": result = getValue(logic[0]) and getValue(logic[2])
        of "OR": result = getValue(logic[0]) or getValue(logic[2])
        of "LSHIFT": result = getValue(logic[0]) shl getValue(logic[2])
        of "RSHIFT": result = getValue(logic[0]) shr getValue(logic[2])
  wire.hasValue = true
  wire.value = result

let firstPart = getValue("a")
echo firstPart

for wire in wires.keys:
  wires[wire].hasValue = false
wires["b"].value = firstPart
wires["b"].hasValue = true

let secondPart = getValue("a")
echo secondPart

File: python/day07.py
from functools import lru_cache
from operator import and_, or_, lshift, rshift


parse_input = lambda data: {(s := line.split(' -> '))[1]: s[0].split() for line in data}

def get_value(wires, wire='a', b=None):
    operations = {'AND': and_, 'OR': or_, 'LSHIFT': lshift, 'RSHIFT': rshift}
    @lru_cache
    def aux(wire):
        try: return int(wire)
        except ValueError: lhs = wires[wire]
        return ( aux(lhs[0])          if len(lhs) == 1 else
                ~aux(lhs[1]) & 0xffff if len(lhs) == 2 else
                operations[lhs[1]](aux(lhs[0]), aux(lhs[2])))
    if b is not None: wires['b'] = [b]
    return aux(wire)


data = open("inputs/07.txt").read().splitlines()
wires = parse_input(data)

print(first := get_value(wires))
print(get_value(wires, b=first))